File: hungry_birds.c
/*
 * hungry birds - an example of multi-thread application using the shared memory
 *                model. A producer/consumer problem.
 *
 * author:        Paolo Rovelli <paolo.rovelli@ericsson.com>
 *
 * to build:	  gcc -g -Wall -lpthread -o hungry_birds hungry_birds.c
 * to run:	  ./hungry_birds
 */

#ifndef _REENTRANT
#define _REENTRANT
#endif
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <pthread.h>
#include <semaphore.h>

#define SHARED 1
#define NUM_BIRDS 50
#define NUM_WORMS 100

void *bird_parent(void *arg);
void *bird_child(void *arg);

static int worms;
static sem_t empty, any;
static pthread_mutex_t lock;

int main(int argc, char *argv[])
{
	pthread_t pid, cid[NUM_WORMS];
	pthread_attr_t attr;
	int i;

	pthread_attr_init(&attr);
	pthread_attr_setscope(&attr, PTHREAD_SCOPE_SYSTEM);

	worms = NUM_WORMS;
	printf("\n + intitial dish %d worms\n\n", worms);
	sem_init(&empty, SHARED, 0);
	sem_init(&any, SHARED, NUM_WORMS);
	pthread_mutex_init(&lock, NULL);

	pthread_create(&pid, &attr, bird_parent, (void *) 0);
	for (i = 0; i < NUM_BIRDS; i++)
		pthread_create(&cid[i], &attr, bird_child, (void *) i);

	pthread_join(pid, NULL);
	for (i = 0; i < NUM_BIRDS; i++)
		pthread_join(cid[i], NULL);

	pthread_exit(NULL);
}

void *bird_parent(void *arg)
{
	int i;

	while (1) {
		sem_wait(&empty);
		pthread_mutex_lock(&lock);
		worms = NUM_WORMS;
		printf(" + bird parent brings %d worms\n\n", worms);
		pthread_mutex_unlock(&lock);
		for (i = 0; i < NUM_WORMS; i++)
			 sem_post(&any);
	}
}

void *bird_child(void *arg)
{
	while (1) {
		sleep(rand()%10);
		sem_wait(&any);
		pthread_mutex_lock(&lock);
		worms--;
		if(worms != 0) {
			printf(" - bird child %d eats (remain worms %d)\n",
				(int)arg, worms);
			pthread_mutex_unlock(&lock);
		} else {
			printf(" - bird child %d eats (remain worms %d)"
				" and screams\n\n", (int)arg, worms);
			pthread_mutex_unlock(&lock);
			sem_post(&empty);
		}
	}
}

File: Java/src/main/java/org/epalrov/hungrybirds/HungryBirds.java
/*
 * HungryBirds.java - example of concurrent multi-thread application
 *
 * Copyright (C) 2013 Paolo Rovelli
 *
 * Author: Paolo Rovelli <paolorovelli@yahoo.it>
 */

package org.epalrov.hungrybirds;

import java.util.concurrent.Semaphore;
 
public class HungryBirds {

    private static final int NUM_BIRDS = 7;
    private static final int NUM_WORMS = 13;

    private int birds;
    private int worms;

    private Semaphore lock;
    private Semaphore empty;
    private Semaphore any;

    public HungryBirds() {
        /* init sync objects */
        lock = new Semaphore(1);
        empty = new Semaphore(0);
        any = new Semaphore(NUM_WORMS);

        /* init shared resources */
        worms = NUM_WORMS;
        System.out.format("\n + intial dish: %d worms" +
            "\n\n", worms);

        /* init and start threads */
        (new Thread(new HungryBirdsParent())).start();
        for (birds = 0; birds < NUM_BIRDS; birds++)
            (new Thread(new HungryBirdsBaby(birds))).start();

    }

    private class HungryBirdsBaby implements Runnable {

        private int bird;

        public HungryBirdsBaby(int bird) {
            this.bird = bird;
        }

        @Override
        public void run() {
            while (true) {
                try {
                    Thread.sleep((int)(Math.random()
                        * (1000*NUM_BIRDS)/NUM_WORMS + 1));
                    any.acquire();
                    lock.acquire();
                    worms--;
                    if (worms != 0) {
                        System.out.format(" - baby bird %d eats" +
                            " (dish: %d worms)\n", bird, worms);
                        lock.release();
                    } else {
                        System.out.format(" - baby bird %d eats" + 
                            " (dish: %d worms)" +
                            " and screams\n\n", bird, worms);
                        lock.release();
                        empty.release();
                    }
                } catch (InterruptedException e) {
                    System.out.println(e);
                } catch (Exception e) {
                    System.out.println(e);
                }
            }            
        }

    }

    private class HungryBirdsParent implements Runnable {

        @Override
        public void run() {
            while(true) {
                try {
                    empty.acquire();
                    lock.acquire();
                    worms = NUM_WORMS;
                    System.out.format(" + parent bird brings" +
                         " %d worms\n\n", worms);
                    lock.release();
                    any.release(worms);
                } catch (InterruptedException e) {
                    System.out.println(e);
                } catch (Exception e) {
                    System.out.println(e);
                }
            }
        }

    }

    public static void main(String[] args) {
        new HungryBirds();
    }

}

